src/api/scenarioData/scenarioData.schema.ts

```diff
@@ -0,0 +1,66 @@
+import { Prop, Schema, SchemaFactory } from '@nestjs/mongoose';
+import { Document, Types } from 'mongoose';
+
+export type ScenarioDataDocument = ScenarioData & Document;
+export interface Variable {
+  name: string;
+  value: any;
+  type?: string;
+}
+
+@Schema()
+export class VariableSchema {
+  @Prop({ required: true, type: String })
+  name: string;
+
+  @Prop({ required: true, type: {} })
+  value: any;
+
+  @Prop({ required: false, type: String, default: '' })
+  type: string;
+}
+
+const VariableModelSchema = SchemaFactory.createForClass(VariableSchema);
+
+// impute the type of the value if not provided
+VariableModelSchema.pre<Variable>('save', function (next) {
+  if (!this.type) {
+    this.type = typeof this.value;
+  }
+  next();
+});
+
+@Schema()
+export class ScenarioData {
+  @Prop({ required: true, description: 'The scenario ID', type: Types.ObjectId, default: () => new Types.ObjectId() })
+  _id: Types.ObjectId;
```

Comment:
Is this needed? I think mongo will just create the _id if you don't define it. Then you don't need "_id: undefined!" in the service file either.

Reply:
Mongo will generate the _id without it being defined no problem, but I run into typescript errors when i try to access the _id property without it being defined on the scenarioData schema, as it asserts that it does not exist. Is there a way around this without defining the _id in the schema?

See lines 220-221 of scenarioData.service.ts for an example of the reference to scenario._id.

Reply:
If you use ScenarioDataDocument instead of ScenarioData, that will have the _id defined.